src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerGenotypingEngine.java

```diff
@@ -168,24 +273,82 @@ public CalledHaplotypes(final List<VariantContext> calls, final Set<Haplotype> c
         mergedVC.getAlternateAlleles().stream().filter(a -> !a.isSymbolic()).forEach(a -> result.put(a, new ArrayList<>()));
 
         for (final Haplotype h : haplotypes) {
-            final VariantContext haplotypeVC = h.getEventMap().get(loc);
-            if (haplotypeVC == null) {    //no events --> this haplotype supports the reference at this locus
-                result.get(ref).add(h);
-            } else {
-                //TODO: this only works if eventsAtThisLoc's and mergedVC's alleles are ordered identically.
-                //TODO: this is in fact the case (see AssemblyBasedCallerUtils::makeMergedVariantContext), but
-                //TODO: that's not good enough
-                for (int n = 0; n < eventsAtThisLoc.size(); n++) {
-                    if (haplotypeVC.hasSameAllelesAs(eventsAtThisLoc.get(n))) {
-                        result.get(mergedVC.getAlternateAllele(n)).add(h);
-                        break;
+
+            final Iterator<VariantContext> spanningEvents = h.getEventMap().getSpanningEvents(loc);
+
+            final boolean hasSpanningEvents = spanningEvents.hasNext();
+
+            while (spanningEvents.hasNext()) {
+                final VariantContext spanningEvent = spanningEvents.next();
+                if (spanningEvent.getStart() == loc) {
+                    // the event starts at the current location
+
+                    if (spanningEvent.getReference().length() == mergedVC.getReference().length()) {
+                        // reference allele lengths are equal; we can just use the spanning event's alt allele
+                        // in the case of GGA mode the spanning event might not match an allele in the mergedVC
+                        if (result.containsKey(spanningEvent.getAlternateAllele(0))) {
+                            result.get(spanningEvent.getAlternateAllele(0)).add(h);
+                        }
+                    } else if (spanningEvent.getReference().length() < mergedVC.getReference().length()) {
+                        // spanning event has shorter ref allele than merged VC; we need to pad out its alt allele
+                        final Map<Allele, Allele> spanningEventAlleleMappingToMergedVc
+                                = GATKVariantContextUtils.createAlleleMapping(mergedVC.getReference(), spanningEvent, new ArrayList<>());
```

Comment:
If you feel like going on a tangent, its' weird that `createAlleleMapping` returns a map even though you're only applying it to a single alt allele.  There should be a method to pad a single alt allele to a longer reference.

Reply:
I think you're right but I'm going to pass on this refactor for now.

Reply:
No problem.